### PR TITLE
Mutable Prototype

### PR DESCRIPTION
**Pull request by fao89**
Following: https://hackmd.io/@pulp/deprecation

### CODE REVIEW

File: pulp_ansible/app/galaxy/v3/views.py [hunk 1]
```diff
@@ -315,3 +342,32 @@ def retrieve(self, request, *args, **kwargs):
...
+        Returns a MutableCollectionMetadata queryset for specified distribution.
+        """
+        return RepositoryVersion.objects.filter(
+            pk=self.get_repository_version(self.kwargs["path"]).pk
+        ).prefetch_related(
+            Prefetch(
+                "collection_memberships",
+                queryset=MutableCollectionMetadata.objects.filter(deprecated=True),
+            )
+        )
```
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:373]: Problem: only displays deprecated for the last RepositoryVersion

File: pulp_ansible/app/urls.py [hunk 2]
```diff
@@ -38,6 +38,11 @@
 ]
 
 v3_urls = [
+    path(
+        "metadata/",
+        views_v3.MutableCollectionMetadataViewset.as_view({"get": "list"}),
```
Comment (fao89) [inline pulp_ansible/app/urls.py:43]: image

File: pulp_ansible/app/galaxy/v3/views.py [hunk 3]
```diff
@@ -108,6 +120,21 @@ class CollectionViewSet(
...
+        queryset = super().filter_queryset(queryset)
+        repo_version = self.get_repository_version(self.kwargs["path"])
+        if not repo_version:
+            return queryset
+
+        deprecated = False
+        if self.request.query_params.get("deprecated", "").lower() in ["true", "yes", "1"]:
+            deprecated = True
+
+        return queryset.filter(collection__mutablecollectionmetadata__deprecated=deprecated)
```
Comment (fao89) [inline pulp_ansible/app/galaxy/v3/views.py:136]: image

File: pulp_ansible/app/migrations/0025_deprecation.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# Generated by Django 2.2.16 on 2020-10-12 17:15
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django_lifecycle.mixins
+import uuid
+
+
+class Migration(migrations.Migration):
```
Comment (newswangerd) [inline pulp_ansible/app/migrations/0025_deprecation.py:9]: There should be a step here to migrate the deprecation status over to `MutableCollectionMetadata` before the deprecated field is removed from the collections table.

File: pulp_ansible/app/models.py [hunk 5]
```diff
@@ -236,6 +235,21 @@ class Meta:
...
+    """
+
+    repository_version = models.ForeignKey(
+        RepositoryVersion, on_delete=models.CASCADE, related_name="collection_memberships"
+    )
+    collection = models.ForeignKey(Collection, on_delete=models.CASCADE)
+    deprecated = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = ("collection", "repository_version")
```
Comment (fao89) [inline pulp_ansible/app/models.py:250]: - Should we make it sparse?
- What about CollectionVersion metadata?

File: pulp_ansible/app/galaxy/v3/views.py (hunk 3 above)
Comment (fao89) [reply pulp_ansible/app/galaxy/v3/views.py:136]: Bring all collections by default, only filter deprecation with query params

File: pulp_ansible/app/urls.py (hunk 2 above)
Comment (fao89) [reply pulp_ansible/app/urls.py:43]: Sync should check for the `last_updated`

File: pulp_ansible/app/tasks/collections.py [hunk 6]
```diff
@@ -386,6 +388,19 @@ def _add_collection_version_level_metadata(data, additional_metadata):
             progress_bar.total = count
             progress_bar.save()
 
+            metadata_downloader = remote.get_downloader(
+                url=get_metadata_url(url), silence_errors_for_response_status_codes={404}
+            )
+            try:
+                mutable_data = parse_metadata(await metadata_downloader.run())
+            except FileNotFoundError:
+                mutable_data = {}
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:397]: Sync should consider `last_updated`

File: pulp_ansible/app/tasks/modify.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
+from pulpcore.plugin.models import Content, RepositoryVersion
+
+from pulp_ansible.app.models import AnsibleRepository, MutableCollectionMetadata
+
+
+def add_and_remove(repository_pk, add_content_units, remove_content_units, base_version_pk=None):
```
Comment (fao89) [inline pulp_ansible/app/tasks/modify.py:6]: Used in synclist - "custom" sync between repositories

**Comment by fao89** [comment]
closing in favor of https://github.com/pulp/pulp_ansible/pull/392

**Pull request closed without merging**